portal/app/farm_updater/sonnen_api.py

```diff
@@ -1,38 +1,90 @@
 import requests
+from django.conf import settings
 
 
-def _get_batteries_status_json():
-    # This method does a request to sonnen api to get status
-    serial1 = '67682'
-    # serial2 = '67670'
-    token = '5db92cf858eebce34af146974f49f4d40ec699b99372546c0af628fb48133f61'
-    url_ini = 'https://core-api.sonnenbatterie.de/proxy/'
-    headers = {'Accept': 'application/vnd.sonnenbatterie.api.core.v1+json', 'Authorization': 'Bearer ' + token}
-    status_endpoint = '/api/v1/status'
+class SonnenApiInterface:
 
-    try:
-        resp = requests.get(url_ini + serial1 + status_endpoint, headers=headers)
-        resp.raise_for_status()
-        data = resp.json()
-        data['batt_id'] = serial1
-        return data
+    def __init__(self, url=settings.SONNEN_URL, token=settings.SONNEN_TOKEN):
+        self.url = url
+        self.token = token
+        self.headers = {'Accept': 'application/vnd.sonnenbatterie.api.core.v1+json', 'Authorization': 'Bearer ' +
+                                                                                                      self.token}
 
-    except requests.exceptions.HTTPError as err:
-        print('Error get_battery_status_json')
-        return None
+    def get_batteries_status_json(self, serial):
+        # This method does a request to sonnen api to get status
+        status_endpoint = '/api/v1/status'
+
+        try:
+            resp = requests.get(self.url + serial + status_endpoint, headers=self.headers)
+            resp.raise_for_status()
+            data = resp.json()
+            data['batt_id'] = serial
+            return data
+
+        except requests.exceptions.HTTPError as err:
+            print('Error get_battery_status_json')
+            return None
+
+    def manual_mode_control(self, serial, mode='charge', value='0'):
+        control_endpoint = '/api/v1/setpoint/'
+        # Checking if system is in off-grid mode
+        voltage = self.get_batteries_status_json().get_status()['Uac']
+
+        if voltage == 0:
+            print('Battery is in off-grid mode... Cannot execute the command')
+            return {}
+
+        else:
+            try:
+                resp = requests.get(self.url + serial + control_endpoint + mode + '/' + value,
+                                    headers=self.headers)
+                resp.raise_for_status()
+
+            except requests.exceptions.HTTPError as err:
+                print(err)
+
+            return resp.json()
 
 
 # This method is used on scheduler.py to pull data in given periodicity
 def update_battery_status():
     from app.models import FarmDevice
-    json = _get_batteries_status_json()
-    print('json: ', json)
-    if json is not None:
+    serial_batt1 = settings.SONNEN_BATT1
```

Comment:
Don't store these in the settings file. If the batteries ever get deleted, this breaks. The code is brittle with this coupling. Get all FarmDevices of type Sonnen filtered by the home id of the farm. Then send whatever you get from our DB to their API and let their API tells us whether these exist or not.

Reply:
see new PR. I still have one "hardcoded" info for the homes. Dont know how to automate this though